Add `server.experimental_parser_recursion_limit` configuration

SimonSapin opened the pull request:
If an operation exceeds the limit, the router returns an error without processing it.

SimonSapin commented:
This still needs tests, but besides that I’m not very happy with the boilerplate for passing the configured limit around. Should we make `Configuration` available in more places?

abernix commented:
Could close #1413? (Just referencing it for posterity, really!)

o0Ignition0o commented on apollo-router/src/spec/query.rs:
```diff
@@ -35,17 +35,12 @@ pub struct Query {
     /// query.
     #[tracing::instrument(skip_all, level = "trace")]
-    pub fn format_response(
+    pub(crate) fn format_response(
```
nice catch!

o0Ignition0o reviewed (commented):
Looking great!

garypen commented on apollo-router/src/configuration/mod.rs:
```diff
@@ -269,11 +269,17 @@ pub(crate) struct Server {
+    /// Experimental limitation of query depth
+    /// default: 4096
+    #[serde(default = "default_operation_depth_limit")]
+    pub(crate) experimental_operation_depth_limit: u32,
```
We should just make this `usize` since we end up passing it `as usize`.

garypen commented on apollo-router/src/query_planner/bridge_query_planner.rs:
```diff
@@ -175,13 +184,16 @@ impl Service<QueryPlannerRequest> for BridgeQueryPlanner {
     async fn get(&self, key: QueryKey) -> Result<QueryPlannerContent, QueryPlannerError> {
-        let selections = self.parse_selections(key.0.clone()).await?;
+        let selections = self
+            .parse_selections(key.0.clone(), self.operation_depth_limit)
```
Since this is a `self` function, I don't think we need to pass this parameter. We can just create a local variable (to move) in the `parse_selections` fn.

garypen commented on apollo-router/src/query_planner/bridge_query_planner.rs:
```diff
@@ -109,7 +118,7 @@ impl BridgeQueryPlanner {
                 usage_reporting,
             } => {
-                let subselections = node.parse_subselections(&*self.schema);
+                let subselections = node.parse_subselections(&*self.schema, operation_depth_limit);
```
Since this is a `self` function, I don't think we need to pass this parameter. We can just access it from `self` in `plan`.

garypen reviewed (changes requested):
Looks very good. A couple of small requests to clean up the code a little.

One more substantial thing: how do we provide a long term track of this to users? My original intention was to do this via prometheus. i.e.: we would have a metric `operatation_max_depth_exceeded` (perhaps?) and that would provide a clean interface for user consumption. Maybe we need an `experimental` metrics section for things like this which are promoted along with the functionality?

lennyburdette commented:
Summarizing a conversation between @garypen and me today:
* Customers will definitely appreciate a metric they can use to page if depth limits are exceeded too frequently.
* In the case of bad actors fuzzing large operations, I'm not sure what other info we could provide to help with this.
* In the case of good actors making mistakes, we're thinking that Studio usage information should be plenty useful. It'll have the operation name, client identifiers, and GraphQL error data, so it should be easy to track down the offending clients.

SimonSapin commented on NEXT_CHANGELOG.md:
We could, but does it really help? And could we have experimental configs in other sections?

garypen replied:
We agreed to defer this to a possible future PR subsequent to API 1.0 discussion.

SimonSapin commented:
Alright, I’ve addressed review comments and this is in a shape that I like better. We’ve discussed with Gary merging this as a first step, then adding a Prometheus counter. I’ve marked this as closing https://github.com/apollographql/router/issues/1268 but leaving https://github.com/apollographql/router/issues/1413 open.

Something that became apparent when writing tests: the relationship between parser recursion and how deep a query looks to humans is not obvious. The test query `{ me { reviews { author { reviews { author { name } } } } } }` needs the limit to be at least 12.

lennyburdette commented:
> The test query { me { reviews { author { reviews { author { name } } } } } } needs the limit to be at least 12.

Interesting! Is that because selection sets are one level and fields are another level? How do fragments play into this? Do inline fragments count as another level even though they don't have material impact on the execution of the operation?

garypen commented:
> > The test query { me { reviews { author { reviews { author { name } } } } } } needs the limit to be at least 12.
> 
> Interesting! Is that because selection sets are one level and fields are another level? How do fragments play into this? Do inline fragments count as another level even though they don't have material impact on the execution of the operation?

I think we really need some input from @lrlna (when she returns from vacation) on this, but I think this is probably due to the difference between "parser implementation" (which is what the recursion limit is really measuring) and "operation depth" (which is what we have re-branded it as). We probably need to understand the difference between these two things before we merge this.

FWIW: I had some difficulty understanding that when I added support for recursion limits to apollo-rs. You can see it took me two tries to get it right: https://github.com/apollographql/apollo-rs/issues?q=assignee%3Agarypen+is%3Aclosed

lennyburdette commented:
Yeah, I think that "semantic" depth limit versus parser depth limit is preferable. I just wrote one up using apollo-compiler (in a private repo) and it's pretty straightforward so if we want to punt on this change I think that's fine.

bnjjj reviewed (approved):
I'm also in favor of having this kind of configuration as @o0Ignition0o mentioned earlier:

```yaml
  server:
    experimental:
      operation_depth_limit: 10
      defer_support: true
      something_else: hello
```

instead of having several fields prefixed by `experimental_` but it's pretty subjective

SimonSapin commented:
As discussed with @garypen I’ve renamed this back to `experimental_parser_recursion_limit` since query depth limiting is gonna be a plugin.

I’ve also added a commit to also apply the limit to schema parsing. It’s less sensitive since the schema is presumably more trusted than queries, but it’s more consistent this way. This required a couple more minor breaking API changes as shown in the changelog.

For the configuration shape I don’t have a strong opinion, I don’t mind making the change if the team agrees (though it could be another PR)

garypen commented on NEXT_CHANGELOG.md:
```diff
@@ -43,6 +43,35 @@ so it should not change depending on the variables or HTTP headers.
 
+### Change `PluggableRouterServiceBuilder` methods ([PR #1437](https://github.com/apollographql/router/pull/1437))
+
+`with_naive_introspection` and `with_defer_support` where two parameter-less methods
```
```suggestion
`with_naive_introspection` and `with_defer_support` were two parameter-less methods
```

garypen commented on NEXT_CHANGELOG.md:
```diff
@@ -43,6 +43,35 @@ so it should not change depending on the variables or HTTP headers.
+
+By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/1437
+
+### The `SchemaKind` enum had its `Instance` variant replaced with a `String` ([PR #1437](https://github.com/apollographql/router/pull/1437))
+
+As the name suggests, the new variant contains a `String` instead of `Box<apollo_router::Schema>`,
+so you no longer need to parse the schema before giving it to the router.
+
+By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/1437
+
+### `Schema` no longer implements `FromStr` ([PR #1437](https://github.com/apollographql/router/pull/1437))
+
+This means that `str.parse::<apollo_router::Schema>()` is no longer available.
+If you still need a parsed `Schema` (see above),
+use `apollo_router::Schema(str, &configuration)` instead.
+To use the default `apollo_router::Configuration` 
+you can call `apollo_router::Schema(str, &Default::default())`.
+
+By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/1437
```
I like these changes, but you probably want more opinions than just mine on these changes...
